csrc/fusion_segmenter.cpp

```diff
@@ -1554,7 +1554,9 @@ std::unique_ptr<Fusion> SegmentedFusion::makeFusion(SegmentedGroup* sg) {
     }
   }
 
-  for (auto out : getAllOutputs(sg)) {
+  // note, we would want to keep output consistent and not artificially drop
+  // duplicates.
+  for (auto out : sg->output_vals) {
```

Comment:
Do we need to do something similar for inputs?

Reply:
We wouldn't want that for inputs. i.e. if we have duplicated inputs, we would want to converge those instead.

Reply:
Hmm, I'm not 100% clear why we would want to have duplicated inputs or outputs, but if that's what the given fusion has, wouldn't we need to preserve that, no matter if it's an input or output?

Reply:
Also, I wonder why we would need to do duplicate filtering for outputs. There should be no duplicated output unless there's a duplication in the original fusion, which should be preserved.

It seems to me that inputs are different. We don't want to artificially duplicate inputs but there could be duplications in the original fusion, so we want to only filter duplications due to segmentation but don't want to filter original duplications.

Reply:
Ha, now I see what you mean... I thought you were referring to the segmentation logic that might introduce multiple use of a single tensor. But I think you meant to refer to c++ API `add_input` instead.

I think you are right then -> using c++ API, people could have done that as well. I'll try to patch that.

Meanwhile, this isn't a real problem with our python API.
We expose `add_output` only, so user have no-way to duplicate inputs. :laughing:

Reply:
>  this isn't a real problem with our python API.

So, does that mean that even if a user gives a single tensor as two aliased inputs, we would consider they are different tensors and not considered duplication?

Reply:
> so we want to only filter duplications due to segmentation but don't want to filter original duplications.

We are only filtering it out from the generated kernel.

In integration, we do have a "matching" signature and return value, where our added alias support would recognize the duplication and just aliasing the corresponding return value. So we aren't paying anything extra for that.

I'm not sure if we have the same alias analysis logic for fusion input... Which we happen to need for another optimization that we were just discussing earlier. I'll open an issue to track this one.

Reply:
> In integration, we do have a "matching" signature and return value, where our added alias support would recognize the duplication and just aliasing the corresponding return value. So we aren't paying anything extra for that.

Does that mean we can filter duplicated outputs from generated kernels as the integration would take care of them?

Reply:
> Does that mean we can filter duplicated outputs from generated kernels as the integration would take care of them?

That's supposed to happen. Let me double check that one, I don't think I have tested it thoroughly with different outputs including duplication in arbitrary order.